src/main/src/main/java/org/geoserver/ows/ProxifyingURLMangler.java

```diff
@@ -1,37 +1,193 @@
 import org.geoserver.config.GeoServer;
 import org.geoserver.platform.GeoServerExtensions;
+import org.vfny.geoserver.util.Requests;
 
 /** A URL mangler that replaces the base URL with the proxied one */
 public class ProxifyingURLMangler implements URLMangler {
 
+    public enum Headers {
+        FORWARDED("X-Forwarded"),
+        FORWARDED_PROTO("X-Forwarded-Proto"),
+        FORWARDED_HOST("X-Forwarded-Host"),
+        FORWARDED_PATH("X-Forwarded-Path"),
+        FORWARDED_FOR("X-Forwarded-For"),
+        HOST("Host");
+
+        private String header;
+
+        Headers(String h) {
+            this.header = h;
+        }
+
+        public String asString() {
+            return header;
+        }
+    }
+
+    public enum ForwardedComponents {
+        FOR("for"),
+        BY("by"),
+        PROTO("proto"),
+        HOST("host"),
+        PATH("path");
+
+        private String comp;
+
+        ForwardedComponents(String c) {
+            this.comp = c;
+        }
+
+        public String asString() {
+            return comp;
+        }
+    }
+
     GeoServer geoServer;
 
+    public static String TEMPLATE_SEPARATOR = " ";
+    public static String TEMPLATE_PREFIX = "${";
+    public static String TEMPLATE_POSTFIX = "}";
+
+    public static Map<String, Pattern> FORWARDED_PATTERNS = new HashMap<String, Pattern>();
+
+    {
+        Arrays.asList(ForwardedComponents.values())
+                .forEach(
+                        (comp) -> {
+                            FORWARDED_PATTERNS.put(
+                                    comp.asString(),
+                                    Pattern.compile(
+                                            String.format(
+                                                    "(.*)%s=([^;^ ]+)(.*)", comp.asString())));
+                        });
+    }
+
     public ProxifyingURLMangler(GeoServer geoServer) {
         this.geoServer = geoServer;
     }
 
     public void mangleURL(
             StringBuilder baseURL, StringBuilder path, Map<String, String> kvp, URLType type) {
 
-        // first check the system property
-        String proxyBase = GeoServerExtensions.getProperty("PROXY_BASE_URL");
-        if (proxyBase == null) {
-            // if no system property fall back to configuration
-            proxyBase = geoServer.getSettings().getProxyBaseUrl();
-        }
+        // first check the system property, then fall back to configuration
+        String proxyBase =
+                (GeoServerExtensions.getProperty(Requests.PROXY_PARAM) != null)
+                        ? GeoServerExtensions.getProperty(Requests.PROXY_PARAM)
+                        : this.geoServer.getSettings().getProxyBaseUrl();
+
+        // Mangles the URL base in different ways based on a flag
+        // (for two reasons: a) speed; b) to make the admin aware of
+        // possible security liabilities)
+        baseURL =
+                (this.geoServer.getGlobal().isUseHeadersProxyURL() == true && proxyBase != null)
```

Comment:
Believe the NPE is here, isUseHeadersProxyURL() return a Boolean, which can be null

Reply:
Yes, I tested it interactively, but not the on the latest GeoTools.

Reply:
Yes, the geoserver field is always not null, this class is instantiated by Spring

Reply:
The NPE has also broken all CITE tests with `java.lang.NullPointerException at org.geoserver.ows.ProxifyingURLMangler.mangleURL(ProxifyingURLMangler.java:93)`
https://build.geoserver.org/job/cite-wms-1.3/1735/consoleFull